### ISSUE TITLE
Adding Clustering via Extended Similarity Metrics

### ISSUE THREAD

**Issue by drroe** [Amber-MD/cpptraj#1049]
In collaboration with @ramirandaq @lexin-chen, expand the cluster analysis capabilities of cpptraj by adding clustering via extended similarity metrics (and more).

Some background reading:

https://link.springer.com/article/10.1186/s13321-021-00505-3

https://link.springer.com/article/10.1007/s10822-022-00444-7

**Comment by drroe** [comment]
> Is this the complementary similarity used to then find medoids and outliers in the trajectory?

Yes - it's equivalent to the `gen_sim_dict` routine from `src/tools/esim_modules.py` in MDANCE.

**Comment by ramirandaq** [comment]
gen_sim_dict will take as an input a set of frames/conformations, and output a number (the extended similarity) for the whole set, not a number for every frame. To calculate the outliers and medoids, the function is calculate_comp_sim (in src/tools/bts.py). The complementary similarity does assign a number to every frame in a set, which can be used to rank the frames from high- to low-density.

**Comment by drroe** [comment]
> gen_sim_dict will take as an input a set of frames/conformations, and output a number (the extended similarity) for the whole set, not a number for every frame. 

Yes, I understand that. Let me be more clear.

The `ExtendedSimilarity::Comparison()` function is most like `gen_sim_dict`. The `ExtendedSimilarity::CalculateCompSim()` function (which is what the `extendedcomp` command, Exec_ExtendedComparison class) is using under the hood is more like `calculate_comp_sim`. Let me know if you have any more questions.
